internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go

```diff
@@ -1200,3 +1202,211 @@ func (s *Server) Memory(ctx context.Context, in *empty.Empty) (reply *machine.Me
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := etcd.NewClientFromControlPlaneIPs(ctx, s.Controller.Runtime().Config().Cluster().CA(), s.Controller.Runtime().Config().Cluster().Endpoint())
+	if err != nil {
+		return nil, err
+	}
+
+	// nolint: errcheck
+	defer client.Close()
+
+	resp, err := client.MemberList(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var id *uint64
+
+	for _, member := range resp.Members {
+		if member.Name == hostname {
+			member := member
+			id = &member.ID
+
+			break
+		}
+	}
+
+	if id == nil {
+		return nil, fmt.Errorf("failed to find %q in list of etcd members", hostname)
+	}
+
+	log.Println("leaving etcd cluster")
+
+	_, err = client.MemberRemove(ctx, *id)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = system.Services(nil).Stop(ctx, "etcd"); err != nil {
+		return nil, err
+	}
+
+	// Once the member is removed, the data is no longer valid.
+	if err = os.RemoveAll(constants.EtcdDataPath); err != nil {
+		return nil, err
+	}
+
+	reply = &machine.EtcdMemberRemoveResponse{
+		Messages: []*machine.EtcdMemberRemove{
+			{},
+		},
+	}
+
+	return reply, nil
+}
+
+// EtcdMoveLeader implements the machine.MachineServer interface.
+//
+// nolint: gocyclo
+func (s *Server) EtcdMoveLeader(ctx context.Context, in *machine.EtcdMoveLeaderRequest) (reply *machine.EtcdMoveLeaderResponse, err error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := etcd.NewClientFromControlPlaneIPs(ctx, s.Controller.Runtime().Config().Cluster().CA(), s.Controller.Runtime().Config().Cluster().Endpoint())
+	if err != nil {
+		return nil, err
+	}
+
+	// nolint: errcheck
+	defer client.Close()
+
+	resp, err := client.MemberList(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// nolint: prealloc
+	var (
+		member *etcdserverpb.Member
+		eps    []string
+	)
+
+	for _, m := range resp.Members {
+		eps = append(eps, m.GetClientURLs()...)
+
+		if m.Name == hostname {
+			member = m
+			break
+		}
+	}
+
+	if member == nil {
+		return nil, fmt.Errorf("failed to find %q in list of etcd members", hostname)
+	}
+
+	var (
+		leader *etcdserverpb.Member
+		// TODO: Can we use 127.0.0.1?
+		leaderEp string
+	)
+
+	if len(eps) == 0 {
+		return nil, fmt.Errorf("no endpoints discovered")
+	}
+
+	for _, ep := range eps {
+		var status *clientv3.StatusResponse
+
+		status, err = client.Status(ctx, ep)
+		if err != nil {
+			return nil, err
+		}
+
+		if status.Leader == status.Header.MemberId {
+			for _, m := range resp.Members {
+				if m.GetID() == status.Header.MemberId {
+					leader = m
+				}
+			}
+
+			leaderEp = ep
+		}
+	}
+
+	if leader == nil {
+		return nil, fmt.Errorf("no leader found")
+	}
+
+	if leader.GetID() == member.GetID() {
+		if leaderEp == "" {
+			return nil, fmt.Errorf("leader endpoint is required")
+		}
+
+		client.SetEndpoints(leaderEp)
+
+		for _, m := range resp.Members {
+			if m.GetID() != member.GetID() {
+				leader = m
+			}
+		}
```

Comment:
should we check for `leader` being nil, as this might happen if we have a single node cluster?

Reply:
Added an explicit check for this at the top of the func.